### DISCUSSION TITLE
How to: represent an entire system with a single statechart schema (i.e. how to merge agents into a single machine)

### DISCUSSION THREAD

**Discussion by MrNovado** [statelyai/xstate#1404] [General]
# Context

Using `xstate` in a component-oriented system allows you to break your entire business logic to however-small (manageable) chunks. Top-level components & their respected machines define abstract phases & transitions between them, while the bottom-level ones deal with more detailed, concrete, and specific cases.

For example, let's make a router:

```js
const routerMachine = createMachine({
  states: { page1, page2, page3, page4, ... }
})

function RouterComponent() {
  const [state, send] = useMachine(routerMachine);
  switch(true) {
    case state.matches('page1'): return <Page1/>;
    case state.matches('.....'): return ...;
    default: return 404;
  }
}
```

And a page:

```js
const someTask = createMachine({
  states: { phase1, phase2, phase3, phase4, ... }
})

function PageComponent() {
  const [state, send] = useMachine(someTask);
  switch(true) {
    case state.matches('phase1'): return ...;
    case state.matches('.....'): return ...;
    default: return <Loader/>;
  }
}
```

Pages may also be however complex as you need them and may consist of several features, each defining their own machines, but that's beside the point.

# The problem

Now that we have several machines at different levels of our system, how can we put them all back together, to make a representation of the entire business?

Why? Use-cases:

- You want to have a single entry-point where the entire business flow is defined to be able to quickly grasp what's your system is doing.
- You want to visually represent the entire-business flow of your system and be able to interact with it (e.g. by dropping it into the visualizer; related issue: https://github.com/davidkpiano/xstate/issues/625), to be able to share the business model to your colleagues/clients and argue about the flow even before ANY actual logic or presentation is done.
- You have a highly decentralized system and every agent/service in it has a `scxml` schema so, in theory, you can put them all together somehow, to visually represent the entire system, interact with it, argue about it, thoroughly test it.

# Solutions (speculation)

The first thing that comes to mind is to manually spread-merge your schemas into a single one. And it seems technically possible (those are just objects, right, should be easy enough?), but you quickly realize:

- How you name your actions, guards, services & activities now matters because they can collide with each other. Several different machines on different levels may define the same action `dispose` for example but may need completely different realizations. We may need a naming convention for this. The issue becomes worse if the machines you are merging are 3rd party (services) and you have no control over their schema-definitions.
- Redefining state transitions is a non-trivial task. In my example above (with the router) when the `someTask` machine transitions to its final state (`phase-final`) -- it may or may not mean a top-level transition should happen (page-switch). Even if we define a specific action for this particular case like `goBack` or `goNext` for example, looking at the `someTask` machine schema in isolation gives us no clue about what this `goBack/Next` action actually entails/refers-to. What's back and what's next? Page2 or 3 or what?

Generally, when you start to put all of your little cubes back together into a single comprehensive cohesive structure you realize that by spreading them apart, by encapsulating them into components in the first place you already lost some extremely valuable glue-esque information, and it's just feels like... you shouldn't have done that?

On the other hand, working with a single machine from the very start quickly becomes unmanageable: team-work is gimped for example. And having no ability to put/merge/incorporate service-machines-schemas into your own feels counter-intuitive. We are working with services / components / agents, but we are defining them in a way that it is non-trivial to extract or merge-back their schema-definitions in order to make a different realization of them or a comprehensive representation of all of them working together?

# TL;DR

I want to be able to create a visual interactive representation of the entire business flow of my system (frontend & backend & all the services included). I want to be able to reason and argue about the entire flow way before any actual / concrete- / integrational- / presentational- work is done. I want to be able to build a schema as a comprehensive self-documented guide which will protect me, my colleagues, and my clients from misinterpreting (forgetting) the flow, which will be the primary knowledge-base for onboarding purposes and be easy to evolve. 

Now... how do I achieve that?

**Comment by UberMouse** [comment]
This is quite similar to what I'm trying to work out at work currently. I've found [this](https://links-lang.org/papers/mscs/Master_Thesis_Thomas_Weber_1450761.pdf) Master thesis on representing web based UIs using actors (and state machines inside the actors) which does have some useful information. However, I'm struggling a bit to map it onto an XState + React system however.

My thinking is that leveraging XStates ability to invoke machines to form a hierarchy is key. This also ensures that visualization works because it's just a matter of recursively visualizing machines and joining them together in one visualization.

Then another system that seems essential is tying a view to a machine, somehow the view gets passed the view tied to any child machines it invokes so it can render their views in the right place. That gives you a mapping from the hierarchy of XState machines to a UI hierarchy. Very unclear how that might be feasible though.

Then you just need an event bus for broadcasting messages to all actors which I figured could be handled by wrapping `useMachine` so active machines get tracked and and `send` could be wrapped to also broadcast the event via the event bus.

But again, struggling to come up with a system that can wire all these bits and pieces together. I haven't been able to come up with or find a better system for managing something like this however. The existing approach we use of having various singleton machines floating around outside of the React context responsible for managing a given slice of the UI doesn't really work that well. It becomes hard to synchronize data between machines and you lose any ability to destroy/create machines based on state transitions higher up the hierarchy (because they are all flat).

**Reply by davidkpiano** [reply]
> My thinking is that leveraging XStates ability to invoke machines to form a hierarchy is key. This also ensures that visualization works because it's just a matter of recursively visualizing machines and joining them together in one visualization.

This thinking is correct - the upcoming visualizer will be able to visualizer a system of interconnected actors.
